pkg/accumulator/accumulator.go

```diff
@@ -0,0 +1,256 @@
+	"github.com/kubescape/go-logger/helpers"
+)
+
+const (
+	ACCUMULATOR_SIZE = 10
+)
+
+const (
+	DROP_EVENT_OCCURRED = "drop event occurred\n"
+)
+
+type containersAccumulator struct {
+	accumulatorDataPerContainer map[string]chan evData.EventData
+	registerMutex               sync.Mutex
+}
+
+type Accumulator struct {
+	accumulatorData                 []map[string][]evData.EventData
+	syncReaderWriterAccumulatorData sync.Mutex
+	firstMapKeysOfAccumulatorData   []string
+	cacheAccumulatorSize            int
+	mainDataChannel                 chan *evData.EventData
+	containersData                  containersAccumulator
+	ebpfEngine                      ebpfeng.EbpfEngineClient
+}
+
+type ContainerAccumulator struct {
+	dataChannel chan evData.EventData
+	containerID string
+}
+
+var myContainerID string
+var accumulatorSingelton *Accumulator
+var lock = &sync.Mutex{}
+
+func createAccumulator() *Accumulator {
+	myContainerID = "111111"
+	accumulatorSingelton = &Accumulator{
+		cacheAccumulatorSize:          ACCUMULATOR_SIZE,
+		accumulatorData:               make([]map[string][]evData.EventData, ACCUMULATOR_SIZE),
+		firstMapKeysOfAccumulatorData: make([]string, ACCUMULATOR_SIZE),
+		mainDataChannel:               make(chan *evData.EventData),
+		containersData: containersAccumulator{
+			accumulatorDataPerContainer: make(map[string]chan evData.EventData),
+		},
+	}
+
+	return accumulatorSingelton
+}
+
+func GetAccumulator() *Accumulator {
+	if accumulatorSingelton == nil {
+		lock.Lock()
+		defer lock.Unlock()
+		if accumulatorSingelton == nil {
+			logger.L().Debug("Creating accumulatorSingelton now.")
+			accumulatorSingelton = createAccumulator()
+		} else {
+			logger.L().Debug("accumulatorSingelton already created.")
+		}
+	} else {
+		logger.L().Debug("accumulatorSingelton already created.")
+	}
+
+	return accumulatorSingelton
+}
+
+func CreateContainerAccumulator(containerID string, dataChannel chan evData.EventData) *ContainerAccumulator {
+	return &ContainerAccumulator{
+		dataChannel: dataChannel,
+		containerID: containerID,
+	}
+}
+
+func (acc *Accumulator) createNewMap(event *evData.EventData, index int) {
+	slice := make([]evData.EventData, 0)
+	m := make(map[string][]evData.EventData)
+	m[event.GetEventContainerID()] = slice
+	acc.accumulatorData[index] = m
+	acc.firstMapKeysOfAccumulatorData[index] = event.GetEventContainerID()
+}
+
+func (acc *Accumulator) getFirstTimestamp() (time.Time, error) {
+	if len(acc.accumulatorData) < 1 {
+		return time.Time{}, fmt.Errorf("getFirstTimestamp failed the slice accumulatorData has no members")
+	}
+	if len(acc.firstMapKeysOfAccumulatorData) < 1 {
+		return time.Time{}, fmt.Errorf("getFirstTimestamp failed the slice firstMapKeysOfAccumulatorData has no members")
+	}
+	if len(acc.accumulatorData[0][acc.firstMapKeysOfAccumulatorData[0]]) < 1 {
+		return time.Time{}, fmt.Errorf("getFirstTimestamp failed the slice acc.accumulatorData[0][acc.firstMapKeysOfAccumulatorData[0] has no members")
```

Comment:
[Required]: This should not be an error message, the error should tell the logic that went wrong and not a name of a variable.